PR 3/5 - Refactor(bookmark): Create BookmarkController (#879)

jose-Diaz1002 opened the pull request:
This PR corresponds to **Task 879** and marks the official migration of the bookmark retrieval endpoint (`GET /users/{userId}/bookmarks`) to the new User Interactions module. This step further decouples the bookmark logic from the core User Identity service, leveraging the dedicated Service and Repository layers created in the previous tasks.

🔹 **Key Changes:**

* **Controller Implementation:** Introduced the `BookmarkController` under the `user-interaction` module.
* **API Contract Preservation:** The existing public endpoint mapping (`GET /users/{userId}/bookmarks`) is preserved, ensuring **zero changes** to the frontend contract.
* **Separation of Concerns:** The controller's sole responsibility is to receive the request and delegate the business logic to the **`BookmarkService`**.
* **New Data Source:** The service layer is invoked to retrieve the `Set<UUID>` of `challengeIds` directly from the **new `bookmarks` collection** (as defined in #877).

These changes ensure the backend is now using the new, scalable architecture for reading bookmark data, while the application's functionality remains unaffected.

## 🎯 Goal: [879] Implement Bookmark Controller

This PR introduces the **Bookmark Controller** layer in the `itachallenge-user` module.

The functionality allows users to **add and remove challenges from their list of bookmarks**, integrating the previously merged Service/Repository changes from the `refactor/878-implement-bookmark-service` branch.

---

## 🎯 Goal: [879] Implement Bookmark Controller

This PR implements the **Bookmark Controller** layer in the `itachallenge-user` module. This provides the API endpoints necessary to manage user challenge bookmarks, utilizing the previously merged Bookmark Service logic.

---

## 🔎 Summary of Core Changes

The review should focus exclusively on the implementation of the new controller and minimal adjustments to the existing user controller.

### 1. 🆕 New API and Test Components (Primary Focus)
* **`controller/userinteraction/bookmark/BookmarkController.java`**: **NEW** - Implements the main API endpoint for bookmark operations (Add/Remove).
* **`dto/userinteraction/bookmark/BookmarkResponseDto.java`**: **NEW** - DTO for the bookmark operation response.
* **`test/java/.../BookmarkControllerTest.java`**: **NEW** - Unit tests for the new `BookmarkController`.

### 2. ✏️ Modified User Components (Your Direct Changes)
* **`controller/UserController.java`**: **MODIFIED** - **The `GET` method was adjusted** to correctly handle or integrate related bookmark functionality.
* **`test/java/.../UserControllerTest.java`**: **MODIFIED** - Tests for the existing `UserController` were adjusted:
    * **Removed tests for standard error responses** (400, 404, 500, etc.) to simplify the test suite.
    * Updated tests to accommodate `UserController.GET` changes.

### 3. ♻️ Inherited Changes (From Merged Branch)

**NOTE TO REVIEWER:** The following modifications to DTOs in the `itachallenge-challenge` module and `SolutionAction.java` were **inherited and merged from `refactor/878-implement-bookmark-service`** (the Service implementation) and were **not directly modified by this branch**.

* **`itachallenge-user/src/main/java/.../SolutionAction.java`**
* **`itachallenge-challenge/src/main/java/.../ChallengeSolvedController.java`**
* **`itachallenge-challenge/src/main/java/.../ChallengeCreateDto.java`** (and related DTO files/tests)
* `CHANGELOG.md`


#

This task is the third of a series of 5:
* ~~#877 Create "Bookmark" documents and repositories (new collection)~~ (Completed by PR #1040)
* ~~#878 Implement "BookmarkService" (the logic)~~ (Completed by PR #1041)
* **#879 Implement "BookmarkController" (getting user's bookmarks from new collection): THIS PR**
* #880 Integration tests + Documentation update (Next step)
* #881 Delete field "bookmarkChallenges" from the UserDocument

---
## ✅ Testing Performed

Bookmarks endpoints have been verified via Postman:

- **Refactored in BookmarkController:** GET
- **Remained in UserController:** POST and DELETE

All endpoints are working correctly. Screenshots attached as evidence.

<img width="1493" height="916" alt="postmanDeleteFalse" src="https://github.com/user-attachments/assets/6fce3c70-49eb-4425-a1f7-4041f3a5f401" />
<img width="1496" height="930" alt="postmanDeleteTrue" src="https://github.com/user-attachments/assets/fabc27d9-0d6e-4fcf-b196-98f776f9d0c2" />
<img width="1492" height="825" alt="postmanAddFalse" src="https://github.com/user-attachments/assets/8282d2b0-31c6-4ef9-ab15-16c04feab00b" />
<img width="1494" height="751" alt="postmanCreateAdd" src="https://github.com/user-attachments/assets/7d6ae754-d8fa-4fcd-93c7-8d268814b843" />
<img width="1486" height="747" alt="postman1" src="https://github.com/user-attachments/assets/a6d8ea38-4a97-415d-8158-a9cf69af7ae2" />

---

- [x] New `BookmarkController` implemented under the `user-interaction` module.
- [x] Existing API contract for `GET /users/{userId}/bookmarks` is maintained without changes.
- [x] The controller correctly delegates bookmark retrieval to `BookmarkService`.
- [ ] Code follows agreed-upon project conventions (naming, formatting, Sonar rules).
- [ ] All unit tests pass locally and in the development environment.
- [ ] SonarQube/SonarCloud analysis passes with no new issues or code smells.

aurelien-darbellay committed:
SolutionDto updated to use messages defined in messages.properties

aurelien-darbellay committed:
Added Test suite to make sure the validation of SolutionDto works

aurelien-darbellay committed:
annotations in ChallengeCreateDto modified to retrieve messages from messages.properties

aurelien-darbellay committed:
added test for ChallengeCreateDto

aurelien-darbellay committed:
modified ResourceDto to retrieve validation messages from messages.properties

aurelien-darbellay committed:
ResourceDtoTest updated to include more validation testing

aurelien-darbellay committed:
ChallengeControllerTest updated to take into account the modification in SolutionDto validation annotations

aurelien-darbellay committed:
fix: remove phantom changes from ChallengeControllerTest

aurelien-darbellay committed:
fix: remove phantom changes ChallengeControllerTest.java

aurelien-darbellay committed:
fix:remove phantom changes in ChallengeControllerTest.java

aurelien-darbellay committed:
fix:remove phantom changes in ChallengeControllerTest

aurelien-darbellay committed:
fix: remove phantom changes in ResourceDtoTest.java

aurelien-darbellay committed:
fix:remove phantom changes in ChallengeControllerTest.java

aurelien-darbellay committed:
fix: remove phantom changes in ChallengeControllerTest.java

aurelien-darbellay committed:
Fix formatting of JSON string in test case

aurelien-darbellay committed:
Fix JSON formatting in provideInvalidEnumValues method

aurelien-darbellay committed:
Fix JSON formatting in provideInvalidEnumValues method

aurelien-darbellay committed:
Fix formatting and spacing in ChallengeControllerTest

itsAidenDev committed:
Created bookmark docs and repo

itsAidenDev committed:
Created bookmark docs and repo

itsAidenDev committed:
Added more testing to BookmarkResponseDtoTest

itsAidenDev committed:
Added more testing to BookmarkResponseDtoTest

itsAidenDev committed:
Reversed changes in UserServiceImpl

itsAidenDev committed:
Reversed changes in UserServiceImpl

itsAidenDev committed:
Created new BookmarkService and BookmarkServiceImpl + testing

itsAidenDev committed:
Fix update in UserServiceImpl and UserServiceImplTest

zohra-b committed:
created enum + tests for deserialization

zohra-b committed:
quitted phantom change

zohra-b committed:
updated changelog

jose-Diaz1002 committed:
feat(user-interaction): Implement BookmarkController for GET /users/{userId}/bookmarks #879

jose-Diaz1002 committed:
feat(user-interaction): Implement BookmarkController for GET /users/{userId}/bookmarks #879

itsAidenDev committed:
BookmarkRepository implemented in BookmarkServiceImpl (testing still needs fixing)

zohra-b committed:
fixed addToBookmarks method

ivilarop committed:
fix(challenge): align solved endpoint to /solved/{challengeId}

zohra-b committed:
fixing tests in userServuceImpl

jose-Diaz1002 committed:
feat(user-interaction): Implement BookmarkControllerTest for GET #879

jose-Diaz1002 committed:
feat(user-interaction): Implement BookmarkControllerTest for GET #879

zohra-b committed:
changed according to mentor's comments

zohra-b committed:
tests fixed

zohra-b commented:
Please add in your PR description the files you created or you modified. This helps reviewers avoiding them to waste time on reviewing 22 files while you just modified 3 o 4

zohra-b reviewed (changes requested):
BookmarkController + BookmarkControllerTest : ok 

You just need to make a rebase from the PR2/5

zohra-b committed:
fixed sonarCloud issues

zohra-b committed:
fixed test in SoltuonActionTest

jose-Diaz1002 commented:
Hi @Zohra-b,

**Done!** I've updated the PR description to include a detailed and categorized summary of the modified files, as you requested.

I have clearly specified the **Core Logic** files (`BookmarkController` and its tests) and also added a note clarifying that the DTO files in the `itachallenge-challenge` module are changes inherited from a merge, not direct modifications by this branch.

Please review the updated description when you have a moment. Thanks for the feedback!

Regards,
Jose Diaz

FlavioKde reviewed (changes requested):
This migration is successful because it maintains the exact same API contract (GET /users/{userId}/bookmarks returns Set<UUID>) that the frontend expects. Well done in that regard!

But :

UserController - Incorrect dependency injection
Problem: You are injecting BookmarkService into UserController but not using it, and this also violates the single responsibility principle.

BookmarkController - Missing UUID validation
Problem: You are not validating that the userId is a valid UUID before passing it to the service.

jose-Diaz1002 committed:
test: stabilize controller tests (#879)

jose-Diaz1002 committed:
test: Resolve merge conflicts in BookmarkControllerTest (#879)

jose-Diaz1002 committed:
feat(user-interaction):Del UserController Se ha eliminado la inyección de BookmarkService #879

jose-Diaz1002 commented:
Hello @FlavioKde ,

Thank you for the thorough review and for confirming that the migration successfully maintains the expected API contract. That's great news!

I've implemented the correction for the architecture issue immediately, as it was a direct violation of the Single Responsibility Principle (SRP).

✅ Correction Applied (SRP Principle)
UserController: The unused BookmarkService dependency has been removed from the class properties and the constructor. This corrects the dependency injection issue and adheres to the SRP.

⏸️ Pending Point: UUID Validation
Regarding the "Missing UUID validation in BookmarkController" (and other controllers):

I agree with the suggestion: I fully recognize that implementing explicit UUID format validation using @Pattern and @Validated on path variables is a critical robustness improvement that would ensure a clean HTTP 400 response on bad input.

Codebase Consistency: However, I've observed that this specific validation is currently missing across several endpoints in the module (FavoriteController, other UserController paths, etc.).

Proposed Scope: To keep this PR focused purely on the migration and the immediate SRP fix, I've decided to not include the UUID validation at this time.

I propose we create a new dedicated task (Jira/GitHub issue) to implement "Uniform UUID Format Validation on all ID Path Variables" across the entire module. This approach ensures the validation is applied consistently, avoids scope creep in this PR, and treats the improvement as a systematic refactor rather than a patch.

Please let me know if you have any comments on this approach or if you'd still prefer the validation to be included here.

Best regards,

jose_Diaz1002

zohra-b committed:
Change JSON property name from 'uuid_favorite' to 'uuid_bookmark'

zohra-b committed:
Change JSON property name from 'uuid_favorite' to 'uuid_bookmark'

zohra-b committed:
Update JsonProperty value for uuid field in tests

zohra-b committed:
Update JsonProperty value for uuid field in tests

zohra-b committed:
Refactor BookmarkDocumentTest by removing redundant tests

Removed multiple tests related to equals, hashCode, and builder methods for BookmarkDocument.

zohra-b committed:
Refactor BookmarkDocumentTest by removing redundant tests

Removed multiple tests related to equals, hashCode, and builder methods for BookmarkDocument.

zohra-b committed:
Remove unused bookmarkDocument2 in tests

Removed unused bookmarkDocument2 initialization from tests.

zohra-b committed:
Remove unused bookmarkDocument2 in tests

Removed unused bookmarkDocument2 initialization from tests.

zohra-b committed:
Remove redundant equality and hashCode tests

Removed multiple tests related to equals, hashCode, and builder methods from BookmarkResponseDtoTest.

zohra-b committed:
Remove redundant equality and hashCode tests

Removed multiple tests related to equals, hashCode, and builder methods from BookmarkResponseDtoTest.

zohra-b committed:
Quitted @data to avoid Sonar request coverage

zohra-b committed:
Quitted @data to avoid Sonar request coverage

zohra-b committed:
Add missing closing brace in BookmarkDocumentTest

zohra-b committed:
Add missing closing brace in BookmarkDocumentTest

zohra-b committed:
fixed Sonar issue 

changed "exists ? // : // " to "exists == Boolean.TRUE ? // : //"

zohra-b committed:
Remove unused import in UserController

Removed unused import for BookmarkService.

zohra-b committed:
Sonar issue : Fix conditional check for user existence in bookmarks

exists ? 
replaced by 
exists == Boolean.TRUE ?

DiegoBalaguer reviewed (approved):
**BookmarkResponseDto**

Use "Data" and ·Getter" when "Data" already includes a "Getter". If you only need a "Getter", it's better to remove "Data" since you're creating many more entry points, and SonicQ will force you to create many more tests.

Great work!!

jose-Diaz1002 commented:
> **MarcadorRespuestaDto**
> 
> Usa "Datos" y "Getter" cuando "Datos" ya incluya un "Getter". Si solo necesitas un "Getter", es mejor eliminar "Datos", ya que estás creando muchos más puntos de entrada y SonicQ te obligará a crear muchas más pruebas.
> 
> ¡¡¡Buen trabajo!!

##  Hi @DiegoBalaguer 

Thank you for your feedback regarding the `BookmarkResponseDto` and the Lombok annotations!

I've checked the `BookmarkController` that I refactored, and it currently doesn't use a `BookmarkResponseDto`. The endpoint returns a `Set<UUID>` directly:

```java
@GetMapping("/users/{userId}/bookmarks")
public Mono<ResponseEntity<Set<UUID>>> getUserBookmarks(@PathVariable String userId) {
    return bookmarkService.getUserBookmarks(userId)
            .map(ResponseEntity::ok);
}

FlavioKde reviewed (approved):
Thanks for the clarification, OK from my end

aperez1haad committed:
Revert "Replace data for getter/setter"

This reverts commit fb0a57fb77879f431e9e22e974d63005c40f4ff3.

aperez1haad committed:
Revert "Replace data for getter/setter"

This reverts commit fb0a57fb77879f431e9e22e974d63005c40f4ff3.

eduzavarce committed:
Refactor: Extract common fields to InteractionDocument and extend in BookmarkDocument and FavoriteDocument

eduzavarce committed:
Refactor: Extract common fields to InteractionDocument and extend in BookmarkDocument and FavoriteDocument

eduzavarce committed:
Remove unused @AllArgsConstructor import from BookmarkDocument

eduzavarce committed:
Remove unused @AllArgsConstructor import from BookmarkDocument

eduzavarce committed:
Remove redundant assertions and add missing test annotation in FavoriteDocumentTest

eduzavarce committed:
Remove redundant assertions and add missing test annotation in FavoriteDocumentTest

eduzavarce committed:
Consolidate and refactor equals, hashCode, and toString tests in BookmarkDocumentTest and FavoriteDocumentTest

eduzavarce committed:
Consolidate and refactor equals, hashCode, and toString tests in BookmarkDocumentTest and FavoriteDocumentTest

zohra-b committed:
updated changelog.md

zohra-b committed:
updated changelog.md

zohra-b committed:
Fix typo in CHANGELOG for InteractionDocument

zohra-b committed:
rebase with develop

zohra-b committed:
Merge branch 'refactor/877-create-bookmarks-docs-and-repo' into refactor/878-implement-bookmark-service

jose-Diaz1002 committed:
Merge branch 'develop' of https://github.com/IT-Academy-BCN/ita-challenges-backend into refactor/879-implement-bookmark-controller

jose-Diaz1002 committed:
Merge branch 'refactor/878-implement-bookmark-service' into refactor/879-implement-bookmark-controller

zohra-b committed:
made InteractionDocument class abstract

zohra-b committed:
quitted AllArgConstructor annotation

zohra-b committed:
turned InteractionDocument fields private (not protected) and turned to add AllArgConstructors for build was unsuccessful without

zohra-b committed:
quitted constructors in favoritedocument and in Bookmarkdocument to use SuperBuilder

zohra-b committed:
modified entire BookmarkDocumentTest class to maintain one only test

zohra-b committed:
quitted no arg / all arg test - coverage still 100

zohra-b committed:
simplified BookmarkResponseDto test : 2 tests remainings

zohra-b committed:
refactor BookmarkdocumentTest

zohra-b committed:
refactored FavoriteDocumentTest : one only test remaining

zohra-b committed:
refactored FavoriteResponseDtoTest : one test, coverage 100%

zohra-b committed:
Merge branch 'refactor/877-create-bookmarks-docs-and-repo' into refactor/878-implement-bookmark-service

zohra-b committed:
fixed a test in UserServiceImpltest

ivilarop committed:
test(userinteraction): simplify document tests and keep coverage

ivilarop committed:
test(userinteraction): simplify document tests and keep coverage

ivilarop committed:
test(userinteraction): simplify document tests and keep coverage

zohra-b committed:
Merge branch 'refactor/877-create-bookmarks-docs-and-repo' into refactor/878-implement-bookmark-service

zohra-b committed:
Merge branch 'refactor/878-implement-bookmark-service' into refactor/879-implement-bookmark-controller

ivilarop commented:
Extensive pull request with changes already merged into develop. We've opened a new, clean pull request: https://github.com/IT-Academy-BCN/ita-challenges-backend/pull/1056
